chainermn/link.py

```diff
@@ -0,0 +1,191 @@
+                        rank_in=1,
+                        rank_out=None)
+
+
+            class Model1(chainermn.MultiNodeChainGroup):
+
+                def __init__(self, comm):
+                    super(Model1, self).__init__(comm)
+                    self.add_link(MLP(2000, 500, 100), rank_in=0, rank_out=0)
+
+
+        ``Model0`` is expected to be on rank=0, and ``Model1`` is expected to
+        be on rank=1. The first ``MLP`` in ``Model0`` will send its outputs
+        to ``Model1``, then ``MLP`` in ``Model1`` will receive it and send
+        its outputs to the second ``MLP`` in ``Model0``.
+
+    Args:
+        comm (chainermn.communicators._base.CommunicatorBase):
+            ChainerMN communicator.
+    """
+
+    def __init__(self, comm):
+        chainer.utils.experimental('chainermn.MultiNodeChainGroup')
+        super(MultiNodeChainGroup, self).__init__()
+        self._comm = comm
+        self._rank_inouts = []
+
+    def add_link(self, link, rank_in=None, rank_out=None):
+        """Register one connected link with its inout rank.
+
+        Args:
+            link (chainer.Link): The link object to be registered.
+            rank_in (int or list):
+                Ranks from which it receives data. If None is specified,
+                the model does not receive from any machines.
+            rank_out (int or list):
+                Ranks to which it sends data. If None is specified,
+                the model will not send to any machine.
+        """
+        super(MultiNodeChainGroup, self).add_link(link)
+        if isinstance(rank_in, int):
+            rank_in = [rank_in]
+        if isinstance(rank_out, int):
+            rank_out = [rank_out]
+
+        assert rank_in is None or not self._comm.rank in rank_in,\
+            "cannot specify self rank for rank_in"
+        assert rank_out is None or not self._comm.rank in rank_out,\
+            "cannot specify self rank for rank_out"
+
+        self._rank_inouts.append((rank_in, rank_out))
+
+    def __call__(self, *inputs):
+        y = None
+        backward_pointer = None
+
+        for f, (rank_in, rank_out) in zip(self._children, self._rank_inouts):
+            x = None
+
+            if rank_in is None:
+                x = f(*inputs)
+            else:
+                # Preprocess: receiving inputs from the other machine.
+                for _rank_in in rank_in:
+                    _x = chainermn.functions.recv(
+                        self._comm,
+                        rank=_rank_in,
+                        backward_pointer=backward_pointer,
+                        device=self._device_id)
+
+                    # Guarantee the backward path to the previous graph
+                    # component to be executed in the last to avoid dead-lock.
+                    if backward_pointer is not None and _x.creator is not None:
+                        _x.creator.rank = -1
+
+                    x = _x if x is None else x + _x
```

Comment:
I don't think it is reasonable design to sum up received arrays when receiving from multiple workers. I think it would be more natural and useful to hand it to `f` as different parameters.